lib/puppet/resource_api/property.rb

```diff
@@ -69,10 +86,40 @@ def rs_value
   # method overloaded only for the :ensure property, add option to check if the
   # rs_value matches is. Only if the class is child of
   # Puppet::ResourceApi::Property.
-  def def_insync?
+  def def_ensure_insync?
     define_singleton_method(:insync?) { |is| rs_value.to_s == is.to_s }
   end
 
+  def def_custom_insync?
+    define_singleton_method(:insync?) do |is|
+      provider    = @referrable_type.my_provider
+      context     = @referrable_type.context
+      should_hash = @resource.rsapi_canonicalized_target_state
+      is_hash     = @resource.rsapi_current_state
+      title       = @resource.rsapi_title
+
+      raise(Puppet::DevError, 'No insync? method defined in the provider; an insync? method must be defined if the custom_insync feature is defined for the type') unless provider.respond_to?(:insync?)
+      provider_insync_result = provider.insync?(context, title, @attribute_name, is_hash, should_hash)
+      # if nil, super and rely on Puppet::Property.insync?
+      # if true/false, return that and use the default change messaging
+      # if string, set value of custom change reporting and treat as false
+      @change_to_s_value = provider_insync_result if provider_insync_result.is_a?(String)
+      return provider_insync_result == true unless provider_insync_result.nil?
+
+      super(is)
```

Comment:
It seems odd to me that the provider would define `insync?` but return `nil` to get the "default behavior", which they would have gotten if they didn't define `insync?` in the first place. Can we just say if you define `insync?`, then the provider has to handle it (which might include calling `super(is)` on the underlying `Puppet::Property`)?

Reply:
The idea for the DevX here was to require as little additional handling in the provider; _only_ add handling for things which need custom insync and otherwise `super` back to the `Puppet::Property.insync?` method. The proposed UX here would require users to _always_ specify the `else` block _unless_ all properties of the resource require custom insync logic, which I suspect would be very rare.

Reply:
Sure that makes sense. So returning true/false means insync handled it, otherwise nil means do the default.